File: src/heroku_command.ts
import * as child from "child_process";
import * as util from "util";
import { ProcessLines } from "./process_lines";

export class HerokuCommand {
  private childProcess: child.ChildProcess;

  constructor(childProcess: child.ChildProcess) {
    this.childProcess = childProcess;
  }

  public static run(command: string, args: string[] = []): HerokuCommand {
    const fullArgs = [
      command,
      ...args,
      "-a",
      process.env.APP_NAME,
    ];
    const childProcess = child.spawn("heroku", fullArgs, {
      stdio: "pipe"
    });
    return new HerokuCommand(childProcess);
  }

  get outputLines(): AsyncIterable<string> {
    return new ProcessLines({
      stdout: this.childProcess.stdout,
      stderr: this.childProcess.stderr,
    });
  }

  get exitCode(): Promise<number> {
    return new Promise((resolve) => {
      this.childProcess.on("close", resolve);
    });
  }
}

File: src/heroku_command.rs
use std::error::Error;
use std::io::{self, BufRead};
use std::process::*;

pub(crate) struct HerokuCommand {
    output: Output,
}

impl HerokuCommand {
    pub(crate) fn run(
        app_name: &str,
        command: &str,
        args: &[&str],
    ) -> Result<Self, Box<dyn Error>> {
        let output = Command::new("heroku")
            .arg(command)
            .args(args)
            .arg("-a")
            .arg(app_name)
            .output()?;
        Ok(Self { output })
    }

    // fn output_lines(&self) -> impl Stream<Item = Result<String, io::Error>>
    pub(crate) fn output_lines(&self) -> impl Iterator<Item = Result<String, io::Error>> + '_ {
        let stdout_lines = self.output.stdout.lines();
        let stderr_lines = self.output.stderr.lines();
        stdout_lines.chain(stderr_lines)
    }

    // async fn status(&self) -> ExitStatus
    pub(crate) fn status(&self) -> ExitStatus {
        self.output.status
    }
}
